controllers/imapmon.go

```diff
@@ -0,0 +1,184 @@
+*		 - Have a DB field "last_login_error" if last login failed
+*		 - DB counter for non-campaign emails that the admin should investigate
+*		 - Add field to User for numner of non-campaign emails reported
+ */
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/gophish/gophish/logger"
+
+	"github.com/glennzw/eazye"
+	"github.com/gophish/gophish/config"
+	"github.com/gophish/gophish/models"
+)
+
+// Pattern for GoPhish emails e.g ?rid=AbC123
+var goPhishRegex = regexp.MustCompile("(\\?rid=[A-Za-z0-9]{7})")
+
+// ImapMonitor is a worker that monitors IMAP serverd for reported campaign emails
+type ImapMonitor struct {
+	cancel    func()
+	reportURL string
+	checkFreq int
+}
+
+// PeriodicIMAPMonitor will periodically check the database for
+func PeriodicIMAPMonitor(ctx context.Context, im *ImapMonitor) {
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			imaps, err := models.GetEnabledIMAPs()
+			if err != nil {
+				log.Error(err)
+				break
+			}
+			var wg sync.WaitGroup
+			wg.Add(len(imaps))
+			for _, i := range imaps {
+				//We launch a goroutine for each IMAP account. In the future should use some more sophisticated worker/queues
+				go checkForNewEmails(i, &wg, im.reportURL)
+			}
+			wg.Wait()
+			time.Sleep(time.Duration(im.checkFreq) * time.Second)
+		}
+	}
+
+}
+
+// NewImapMonitor returns a new instanfe of the ImapMonitor
+//func NewImapMonitor(configP config.PhishServer, configA *config.Config) *ImapMonitor {
+func NewImapMonitor(config *config.Config) *ImapMonitor {
+	// Make sure database connection exists. Not sure why I have to do this here, but
+	//  otherwise db is <nil> when calling models.GetEnabledIMAPs()
+	err := models.Setup(config)
+	if err != nil {
+		log.Fatal(err)
+	}
+	reportURL := "http://" + config.PhishConf.ListenURL
+	if config.PhishConf.UseTLS {
+		reportURL = "https://" + config.PhishConf.ListenURL
+	}
+
+	im := &ImapMonitor{
+		reportURL: reportURL,
+		checkFreq: config.IMAPFreq,
+	}
+	return im
+}
+
+// Start launches the IMAP campaign monitor
+func (im *ImapMonitor) Start() error {
+	log.Info("Starting IMAP monitor")
+	ctx, cancel := context.WithCancel(context.Background()) //ctx is the derivedContext
+	im.cancel = cancel
+	go PeriodicIMAPMonitor(ctx, im)
+	return nil
+}
+
+// Shutdown attempts to gracefully shutdown the IMAP monitor.
+func (im *ImapMonitor) Shutdown() error {
+	log.Info("Shutting down IMAP monitor")
+	im.cancel()
+	return nil
+}
+
+// checkForNewEmails logs into an IMAP account and checks unread emails
+//  for the rid campaign identifier.
+func checkForNewEmails(im models.IMAP, wg *sync.WaitGroup, reportURL string) {
+	defer wg.Done()
+	mailSettings := eazye.MailboxInfo{
+		Host:   im.Host,
+		TLS:    im.TLS,
+		User:   im.Username,
+		Pwd:    im.Password,
+		Folder: im.Folder}
+
+	//Make sure server ends with a trailing slash
```

Comment:
Small nits - Go typically asks that comments have a space after the slashes. I'm not sure if `gofmt` (or better `goimports`) does this for you but you'll want to make sure that you're using those, mainly because they're super useful 😄

Reply:
Done. There are a few more I found in other Go files too in case you want to change them:

```
grep --include \*.go "[^:]//\S" -ir *

auth/auth.go:	//If we've made it here, we should have a valid user stored in u
auth/auth.go:	//Let's check the password
controllers/route.go:		//Attempt to login
controllers/route.go:		//If we've logged in, save the session and redirect to the dashboard
controllers/api/page.go:	//POST: Create a new page and return it as JSON
controllers/api/util.go:		//default message body
controllers/api/template.go:	//POST: Create a new template and return it as JSON
controllers/api/smtp.go:	//POST: Create a new SMTP and return it as JSON
controllers/api/group.go:	//POST: Create a new group and return it as JSON
controllers/api/campaign.go:	//POST: Create a new campaign and return it as JSON
middleware/session.go:	[]byte(securecookie.GenerateRandomKey(64)), //Signing key
models/campaign_test.go:	//problem
models/models.go:			Hash:     "$2a$10$IYkPp0.QsM81lYYPrQx6W.U6oQGw7wMpozrKhKAHUBVL4mkm/EvAS", //gophish
```